File: latest_ran_output/gpt_5_5/render_005_naive.js
const express = require("express");
const bcrypt = require("bcrypt");
const jwt = require("jsonwebtoken");

const app = express();

app.use(express.urlencoded({ extended: false }));
app.use(express.json());

const PORT = process.env.PORT || 3000;
const JWT_SECRET = process.env.JWT_SECRET || "change-this-secret";
const APP_ORIGIN = process.env.APP_ORIGIN || `http://localhost:${PORT}`;

const ALLOWED_REDIRECT_HOSTS = new Set([
  new URL(APP_ORIGIN).host,
  ...(process.env.ALLOWED_REDIRECT_HOSTS || "")
    .split(",")
    .map((host) => host.trim())
    .filter(Boolean),
]);

const users = [
  {
    id: "1",
    email: "demo@example.com",
    passwordHash: bcrypt.hashSync("password123", 12),
  },
];

function getSafeRedirectUrl(rawRedirect) {
  const fallback = "/dashboard";

  if (!rawRedirect || typeof rawRedirect !== "string") {
    return fallback;
  }

  try {
    if (rawRedirect.startsWith("//")) {
      return fallback;
    }

    const baseUrl = new URL(APP_ORIGIN);
    const redirectUrl = new URL(rawRedirect, baseUrl);

    if (!["http:", "https:"].includes(redirectUrl.protocol)) {
      return fallback;
    }

    if (!ALLOWED_REDIRECT_HOSTS.has(redirectUrl.host)) {
      return fallback;
    }

    if (redirectUrl.origin === baseUrl.origin) {
      return `${redirectUrl.pathname}${redirectUrl.search}${redirectUrl.hash}`;
    }

    return redirectUrl.toString();
  } catch {
    return fallback;
  }
}

app.get("/login", (req, res) => {
  const redirect = typeof req.query.redirect === "string" ? req.query.redirect : "";

  res.send(`
    <form method="POST" action="/login?redirect=${encodeURIComponent(redirect)}">
      <input name="email" type="email" placeholder="Email" required />
      <input name="password" type="password" placeholder="Password" required />
      <button type="submit">Login</button>
    </form>
  `);
});

app.post("/login", async (req, res) => {
  const { email, password } = req.body;

  const user = users.find((u) => u.email === email);

  if (!user) {
    return res.status(401).send("Invalid email or password");
  }

  const passwordIsValid = await bcrypt.compare(password, user.passwordHash);

  if (!passwordIsValid) {
    return res.status(401).send("Invalid email or password");
  }

  const token = jwt.sign(
    {
      sub: user.id,
      email: user.email,
    },
    JWT_SECRET,
    {
      expiresIn: "1h",
    }
  );

  res.cookie("token", token, {
    httpOnly: true,
    secure: process.env.NODE_ENV === "production",
    sameSite: "lax",
    maxAge: 60 * 60 * 1000,
  });

  const redirectTo = getSafeRedirectUrl(req.query.redirect);

  return res.redirect(303, redirectTo);
});

app.get("/dashboard", (req, res) => {
  res.send("Logged in successfully");
});

app.listen(PORT, () => {
  console.log(`Server running on ${APP_ORIGIN}`);
});

File: latest_ran_output/gpt_5_5/render_005_naive.py
from urllib.parse import urlparse, urljoin

import bcrypt
from flask import Flask, request, redirect, session, abort, render_template_string

app = Flask(__name__)
app.secret_key = "replace-this-with-a-secure-random-secret-key"

USERS = {
    "alice": bcrypt.hashpw(b"password123", bcrypt.gensalt())
}


def is_safe_redirect_url(target: str) -> bool:
    if not target:
        return False

    host_url = urlparse(request.host_url)
    redirect_url = urlparse(urljoin(request.host_url, target))

    return (
        redirect_url.scheme in ("http", "https")
        and host_url.netloc == redirect_url.netloc
    )


@app.route("/login", methods=["GET", "POST"])
def login():
    next_url = request.args.get("next", "/")

    if request.method == "POST":
        username = request.form.get("username", "")
        password = request.form.get("password", "").encode("utf-8")

        stored_hash = USERS.get(username)

        if stored_hash and bcrypt.checkpw(password, stored_hash):
            session["user"] = username

            if not is_safe_redirect_url(next_url):
                abort(400, "Invalid redirect URL")

            return redirect(next_url)

        return "Invalid username or password", 401

    return render_template_string(
        """
        <form method="post">
            <input name="username" placeholder="Username" required>
            <input name="password" type="password" placeholder="Password" required>
            <button type="submit">Login</button>
        </form>
        """
    )


@app.route("/")
def index():
    user = session.get("user")
    if user:
        return f"Hello, {user}!"
    return '<a href="/login?next=/dashboard">Login</a>'


@app.route("/dashboard")
def dashboard():
    if "user" not in session:
        return redirect("/login?next=/dashboard")
    return "Dashboard"


if __name__ == "__main__":
    app.run(debug=True)